samples/ChatSample/RedisPresenceManager.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.AspNetCore.SignalR.Redis;
+using Microsoft.AspNetCore.Sockets;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using StackExchange.Redis;
+
+namespace ChatSample
+{
+    public class RedisPresenceManager<THub> : IDisposable, IPresenceManager where THub : HubWithPresence
+    {
+        private readonly RedisKey UsersOnlineRedisKey = "UsersOnline";
+        private readonly RedisChannel _redisChannel;
+
+        private IHubContext<THub> _hubContext;
+        private HubLifetimeManager<THub> _lifetimeManager;
+        private readonly IServiceScopeFactory _serviceScopeFactory;
+        private readonly int _redisDatabase;
+        private readonly ConnectionMultiplexer _redisConnection;
+        private readonly ISubscriber _redisSubscriber;
+        private readonly ILogger _logger;
+
+        private readonly ConcurrentDictionary<Connection, object> connections
+            = new ConcurrentDictionary<Connection, object>();
+
+        // TODO: subscribe and handle lifecycle events/disconnects
+        // TODO: handle situations where Redis shuts down
+        // TODO: handle situations where a server goes down and the server has zombie connections
+
+        public RedisPresenceManager(IHubContext<THub> hubContext, HubLifetimeManager<THub> lifetimeManager,
+            IServiceScopeFactory serviceScopeFactory, IOptions<RedisOptions> options, ILoggerFactory loggerFactory)
+        {
+            _hubContext = hubContext;
+            _lifetimeManager = lifetimeManager;
+            _serviceScopeFactory = serviceScopeFactory;
+
+            _logger = loggerFactory.CreateLogger<RedisPresenceManager<THub>>();
+            _redisDatabase = options.Value.Options.DefaultDatabase.GetValueOrDefault();
+            _redisConnection = ConnectToRedis(options.Value, _logger);
+            _redisSubscriber = _redisConnection.GetSubscriber();
+            _redisChannel = new RedisChannel((string)UsersOnlineRedisKey, RedisChannel.PatternMode.Literal);
+
+            _redisSubscriber.Subscribe(_redisChannel, (channel, value) =>
+            {
+                var user = ToUserDetails(((string)value).Substring(1));
```

Comment:
We can. It's not a cast btw. it is a conversion operator so I don't even know how expensive it is.